File: src/core/parallel-executor.ts
/**
 * ParallelExecutor for RWM Puppet Master
 *
 * Coordinates parallel execution of subtasks using git worktrees for isolation.
 * Uses dependency analysis to determine execution levels and a semaphore
 * to limit concurrent executions.
 *
 * See BUILD_QUEUE_IMPROVEMENTS.md P2-T01 for requirements.
 */

import type { TierNode } from './tier-node.js';
import type { ExecutionEngine, IterationContext, IterationResult } from './execution-engine.js';
import type { PlatformRunnerContract } from '../types/platforms.js';
import type { WorktreeManager, MergeResult, WorktreeInfo } from '../git/worktree-manager.js';
import type { EventBus } from '../logging/event-bus.js';
import {
  buildDependencyGraph,
  getReadySubtasks,
  validateDependencies,
  type DependencyGraph,
} from './dependency-analyzer.js';

/**
 * Result of a parallel execution run
 */
export interface ParallelExecutionResult {
  /** Overall success (all subtasks passed) */
  success: boolean;
  /** Individual results per subtask */
  results: Map<string, SubtaskExecutionResult>;
  /** Subtasks that had merge conflicts */
  conflictSubtasks: string[];
  /** Total execution time in milliseconds */
  totalDurationMs: number;
  /** Maximum concurrency actually used */
  maxConcurrencyUsed: number;
}

/**
 * Result for a single subtask execution
 */
export interface SubtaskExecutionResult {
  /** Subtask ID */
  subtaskId: string;
  /** Whether execution succeeded */
  success: boolean;
  /** Iteration result (if executed) */
  iterationResult?: IterationResult;
  /** Merge result (if merged) */
  mergeResult?: MergeResult;
  /** Worktree info (if created) */
  worktreeInfo?: WorktreeInfo;
  /** Execution time in milliseconds */
  durationMs: number;
  /** Error message if failed */
  error?: string;
  /** Execution level in dependency graph */
  level: number;
}

/**
 * Configuration for ParallelExecutor
 */
export interface ParallelExecutorConfig {
  /** Maximum concurrent executions (default: 3) */
  maxConcurrency: number;
  /** Whether to continue execution after a subtask fails (default: false) */
  continueOnFailure: boolean;
  /** Whether to merge results back to main branch (default: true) */
  mergeResults: boolean;
  /** Target branch for merging (default: current branch) */
  targetBranch?: string;
}

/**
 * Default configuration
 */
const DEFAULT_CONFIG: ParallelExecutorConfig = {
  maxConcurrency: 3,
  continueOnFailure: false,
  mergeResults: true,
};

/**
 * Simple semaphore for limiting concurrency
 */
class Semaphore {
  private currentCount: number = 0;
  private waitQueue: Array<() => void> = [];

  constructor(private readonly maxCount: number) {}

  async acquire(): Promise<void> {
    if (this.currentCount < this.maxCount) {
      this.currentCount++;
      return;
    }

    // Wait for a slot to become available
    return new Promise((resolve) => {
      this.waitQueue.push(() => {
        this.currentCount++;
        resolve();
      });
    });
  }

  release(): void {
    this.currentCount--;
    const next = this.waitQueue.shift();
    if (next) {
      next();
    }
  }

  getCurrentCount(): number {
    return this.currentCount;
  }
}

/**
 * ParallelExecutor coordinates parallel subtask execution
 */
export class ParallelExecutor {
  private readonly config: ParallelExecutorConfig;
  private readonly semaphore: Semaphore;

  constructor(
    private readonly worktreeManager: WorktreeManager,
    private readonly executionEngine: ExecutionEngine,
    private readonly eventBus: EventBus | null,
    config?: Partial<ParallelExecutorConfig>
  ) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.semaphore = new Semaphore(this.config.maxConcurrency);
  }

  /**
   * Execute subtasks in parallel respecting dependencies
   * @param subtasks - Array of subtask TierNodes to execute
   * @param contextBuilder - Function to build IterationContext for each subtask
   * @param runnerProvider - Function to get platform runner for each subtask
   * @returns ParallelExecutionResult with all outcomes
   */
  async executeParallel(
    subtasks: TierNode[],
    contextBuilder: (subtask: TierNode, worktreePath: string) => Promise<IterationContext>,
    runnerProvider: (subtask: TierNode) => PlatformRunnerContract
  ): Promise<ParallelExecutionResult> {
    const startTime = Date.now();
    const results = new Map<string, SubtaskExecutionResult>();
    const conflictSubtasks: string[] = [];
    let maxConcurrencyUsed = 0;

    // Validate dependencies first
    const validation = validateDependencies(subtasks);
    if (!validation.isValid) {
      throw new Error(`Invalid dependencies: ${validation.errors.join('; ')}`);
    }

    // Build dependency graph
    const graph = buildDependencyGraph(subtasks);

    // Emit parallel execution started event
    this.emitEvent('parallel_execution_started', {
      totalSubtasks: subtasks.length,
      levels: graph.levels.length,
      maxConcurrency: this.config.maxConcurrency,
    });

    // Track execution state
    const completedIds = new Set<string>();
    const inProgressIds = new Set<string>();
    const failedIds = new Set<string>();

    // Process all subtasks
    let allSuccess = true;
    let levelIndex = 0;

    for (const level of graph.levels) {
      // Get subtask IDs at this level
      const levelSubtaskIds = new Set(level.map(n => n.id));

      // Execute all subtasks in this level in parallel
      const levelPromises = level.map(async (subtask) => {
        // Wait for semaphore slot
        await this.semaphore.acquire();
        inProgressIds.add(subtask.id);

        // Track max concurrency
        const currentConcurrency = this.semaphore.getCurrentCount();
        maxConcurrencyUsed = Math.max(maxConcurrencyUsed, currentConcurrency);

        try {
          const result = await this.executeSubtaskInWorktree(
            subtask,
            contextBuilder,
            runnerProvider,
            levelIndex
          );
          results.set(subtask.id, result);

          if (result.success) {
            completedIds.add(subtask.id);

            // Handle merge if configured
            if (this.config.mergeResults && result.worktreeInfo) {
              const mergeResult = await this.worktreeManager.mergeWorktree(
                subtask.id,
                this.config.targetBranch
              );
              result.mergeResult = mergeResult;

              if (!mergeResult.success) {
                result.success = false;
                result.error = `Merge conflict: ${mergeResult.conflictFiles.join(', ')}`;
                conflictSubtasks.push(subtask.id);
                failedIds.add(subtask.id);
                allSuccess = false;
              }
            }
          } else {
            failedIds.add(subtask.id);
            allSuccess = false;
          }

          // Emit subtask completed event
          this.emitEvent('parallel_subtask_completed', {
            subtaskId: subtask.id,
            success: result.success,
            level: levelIndex,
            durationMs: result.durationMs,
          });

        } catch (error) {
          const errorMsg = error instanceof Error ? error.message : String(error);
          results.set(subtask.id, {
            subtaskId: subtask.id,
            success: false,
            durationMs: Date.now() - startTime,
            error: errorMsg,
            level: levelIndex,
          });
          failedIds.add(subtask.id);
          allSuccess = false;

          this.emitEvent('parallel_subtask_error', {
            subtaskId: subtask.id,
            error: errorMsg,
            level: levelIndex,
          });

        } finally {
          inProgressIds.delete(subtask.id);
          this.semaphore.release();
        }
      });

      // Wait for all subtasks in this level to complete
      await Promise.all(levelPromises);

      // Check if we should continue after failures
      if (!this.config.continueOnFailure && failedIds.size > 0) {
        // Stop execution - don't process remaining levels
        break;
      }

      levelIndex++;
    }

    const totalDurationMs = Date.now() - startTime;

    // Emit parallel execution completed event
    this.emitEvent('parallel_execution_completed', {
      success: allSuccess,
      totalSubtasks: subtasks.length,
      completedSubtasks: completedIds.size,
      failedSubtasks: failedIds.size,
      conflictSubtasks: conflictSubtasks.length,
      totalDurationMs,
      maxConcurrencyUsed,
    });

    return {
      success: allSuccess && conflictSubtasks.length === 0,
      results,
      conflictSubtasks,
      totalDurationMs,
      maxConcurrencyUsed,
    };
  }

  /**
   * Execute a single subtask in an isolated worktree
   */
  private async executeSubtaskInWorktree(
    subtask: TierNode,
    contextBuilder: (subtask: TierNode, worktreePath: string) => Promise<IterationContext>,
    runnerProvider: (subtask: TierNode) => PlatformRunnerContract,
    level: number
  ): Promise<SubtaskExecutionResult> {
    const startTime = Date.now();
    const agentId = subtask.id;

    // Emit worktree creation event
    this.emitEvent('worktree_creating', { agentId, subtaskId: subtask.id });

    // Create worktree
    let worktreePath: string;
    try {
      worktreePath = await this.worktreeManager.createWorktree(agentId);
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      return {
        subtaskId: subtask.id,
        success: false,
        durationMs: Date.now() - startTime,
        error: `Failed to create worktree: ${errorMsg}`,
        level,
      };
    }

    const worktreeInfo = this.worktreeManager.getWorktree(agentId);

    // Emit worktree created event
    this.emitEvent('worktree_created', {
      agentId,
      subtaskId: subtask.id,
      path: worktreePath,
      branch: worktreeInfo?.branch,
    });

    try {
      // Build iteration context with worktree path
      const context = await contextBuilder(subtask, worktreePath);

      // Get platform runner
      const runner = runnerProvider(subtask);

      // Execute iteration in worktree
      const iterationResult = await this.executionEngine.spawnIteration(context, runner);

      return {
        subtaskId: subtask.id,
        success: iterationResult.success && iterationResult.completionSignal === 'COMPLETE',
        iterationResult,
        worktreeInfo,
        durationMs: Date.now() - startTime,
        level,
      };

    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      return {
        subtaskId: subtask.id,
        success: false,
        worktreeInfo,
        durationMs: Date.now() - startTime,
        error: `Execution failed: ${errorMsg}`,
        level,
      };

    } finally {
      // Clean up worktree on failure or if merge not needed
      if (!this.config.mergeResults) {
        await this.worktreeManager.destroyWorktree(agentId, true).catch(() => {
          // Ignore cleanup errors
        });

        this.emitEvent('worktree_destroyed', {
          agentId,
          subtaskId: subtask.id,
        });
      }
    }
  }

  /**
   * Execute subtasks sequentially (fallback mode)
   * Used when parallel execution is disabled or fails
   */
  async executeSequential(
    subtasks: TierNode[],
    contextBuilder: (subtask: TierNode, worktreePath: string) => Promise<IterationContext>,
    runnerProvider: (subtask: TierNode) => PlatformRunnerContract,
    projectPath: string
  ): Promise<ParallelExecutionResult> {
    const startTime = Date.now();
    const results = new Map<string, SubtaskExecutionResult>();

    // Sort by dependencies if any
    const validation = validateDependencies(subtasks);
    if (!validation.isValid) {
      throw new Error(`Invalid dependencies: ${validation.errors.join('; ')}`);
    }

    let graph: DependencyGraph;
    try {
      graph = buildDependencyGraph(subtasks);
    } catch {
      // If graph building fails, use original order
      graph = { levels: [subtasks], nodes: new Map(), hasDependencies: false, hasCycles: false };
    }

    const sortedSubtasks = graph.levels.flat();
    let allSuccess = true;

    for (let i = 0; i < sortedSubtasks.length; i++) {
      const subtask = sortedSubtasks[i];
      const subtaskStart = Date.now();

      try {
        // Build context with main project path (no worktree)
        const context = await contextBuilder(subtask, projectPath);
        const runner = runnerProvider(subtask);

        // Execute
        const iterationResult = await this.executionEngine.spawnIteration(context, runner);
        const success = iterationResult.success && iterationResult.completionSignal === 'COMPLETE';

        results.set(subtask.id, {
          subtaskId: subtask.id,
          success,
          iterationResult,
          durationMs: Date.now() - subtaskStart,
          level: graph.nodes.get(subtask.id)?.level ?? i,
        });

        if (!success) {
          allSuccess = false;
          if (!this.config.continueOnFailure) {
            break;
          }
        }

      } catch (error) {
        const errorMsg = error instanceof Error ? error.message : String(error);
        results.set(subtask.id, {
          subtaskId: subtask.id,
          success: false,
          durationMs: Date.now() - subtaskStart,
          error: errorMsg,
          level: graph.nodes.get(subtask.id)?.level ?? i,
        });
        allSuccess = false;

        if (!this.config.continueOnFailure) {
          break;
        }
      }
    }

    return {
      success: allSuccess,
      results,
      conflictSubtasks: [],
      totalDurationMs: Date.now() - startTime,
      maxConcurrencyUsed: 1,
    };
  }

  /**
   * Get current configuration
   */
  getConfig(): ParallelExecutorConfig {
    return { ...this.config };
  }

  /**
   * Check if parallel execution is available
   * (worktree manager must be properly initialized)
   */
  isAvailable(): boolean {
    return this.worktreeManager !== null && this.executionEngine !== null;
  }

  /**
   * Emit event to EventBus if available
   */
  private emitEvent(
    type: 'parallel_execution_started',
    data: { totalSubtasks: number; levels: number; maxConcurrency: number }
  ): void;
  private emitEvent(
    type: 'parallel_execution_completed',
    data: { success: boolean; totalSubtasks: number; completedSubtasks: number; failedSubtasks: number; conflictSubtasks: number; totalDurationMs: number; maxConcurrencyUsed: number }
  ): void;
  private emitEvent(
    type: 'parallel_subtask_completed',
    data: { subtaskId: string; success: boolean; level: number; durationMs: number }
  ): void;
  private emitEvent(
    type: 'parallel_subtask_error',
    data: { subtaskId: string; error: string; level: number }
  ): void;
  private emitEvent(
    type: 'worktree_creating',
    data: { agentId: string; subtaskId: string }
  ): void;
  private emitEvent(
    type: 'worktree_created',
    data: { agentId: string; subtaskId: string; path: string; branch?: string }
  ): void;
  private emitEvent(
    type: 'worktree_destroyed',
    data: { agentId: string; subtaskId: string }
  ): void;
  private emitEvent(type: string, data: Record<string, unknown>): void {
    if (this.eventBus) {
      // Type assertion is safe because overloads ensure correct data for each type
      this.eventBus.emit({ type, ...data } as Parameters<typeof this.eventBus.emit>[0]);
    }
  }
}

/**
 * Factory function to create ParallelExecutor
 */
export function createParallelExecutor(
  worktreeManager: WorktreeManager,
  executionEngine: ExecutionEngine,
  eventBus: EventBus | null,
  config?: Partial<ParallelExecutorConfig>
): ParallelExecutor {
  return new ParallelExecutor(worktreeManager, executionEngine, eventBus, config);
}

File: puppet-master-rs/src/core/parallel_executor.rs
//! Parallel subtask executor with dependency awareness
//!
//! Coordinates parallel execution of subtasks with:
//! - Respect for dependency ordering
//! - Configurable concurrency limits
//! - Per-task result tracking
//! - Partial failure handling

use anyhow::{Context, Result};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::Semaphore;
use tokio::task::JoinHandle;

use super::dependency_analyzer::DependencyAnalyzer;

/// Configuration for parallel executor
#[derive(Debug, Clone)]
pub struct ParallelExecutorConfig {
    /// Maximum concurrent executions
    pub max_concurrent: usize,
    /// Whether to continue after a subtask fails
    pub continue_on_failure: bool,
    /// Timeout per task in seconds (0 = no timeout)
    pub task_timeout_secs: u64,
}

impl Default for ParallelExecutorConfig {
    fn default() -> Self {
        Self {
            max_concurrent: 3,
            continue_on_failure: false,
            task_timeout_secs: 3600, // 1 hour default
        }
    }
}

/// Result of a single subtask execution
#[derive(Debug, Clone)]
pub struct SubtaskResult {
    /// Subtask ID
    pub id: String,
    /// Whether execution succeeded
    pub success: bool,
    /// Execution level in dependency graph
    pub level: usize,
    /// Execution duration in milliseconds
    pub duration_ms: u64,
    /// Error message if failed
    pub error: Option<String>,
    /// Optional output data
    pub output: Option<String>,
}

/// Result of parallel execution
#[derive(Debug, Clone)]
pub struct ParallelExecutionResult {
    /// Overall success (all subtasks passed)
    pub success: bool,
    /// Individual results per subtask
    pub results: HashMap<String, SubtaskResult>,
    /// Total execution duration in milliseconds
    pub total_duration_ms: u64,
    /// Maximum concurrency actually used
    pub max_concurrency_used: usize,
    /// Number of subtasks completed
    pub completed_count: usize,
    /// Number of subtasks failed
    pub failed_count: usize,
}

/// Parallel executor for running subtasks concurrently
pub struct ParallelExecutor {
    config: ParallelExecutorConfig,
    analyzer: DependencyAnalyzer,
}

impl ParallelExecutor {
    /// Create new parallel executor
    pub fn new(config: ParallelExecutorConfig) -> Self {
        Self {
            config,
            analyzer: DependencyAnalyzer::new(),
        }
    }

    /// Create with default configuration
    pub fn with_defaults() -> Self {
        Self::new(ParallelExecutorConfig::default())
    }

    /// Execute subtasks in parallel respecting dependencies
    ///
    /// # Arguments
    /// * `subtasks` - Vector of (subtask_id, dependencies)
    /// * `executor` - Async function to execute each subtask
    ///
    /// # Returns
    /// ParallelExecutionResult with all outcomes
    pub async fn execute<F, Fut>(
        &self,
        subtasks: Vec<(String, Vec<String>)>,
        executor: F,
    ) -> Result<ParallelExecutionResult>
    where
        F: Fn(String) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<String>> + Send + 'static,
    {
        let start_time = std::time::Instant::now();

        // Build dependency graph
        let graph = self
            .analyzer
            .build_graph(subtasks.clone())
            .context("Failed to build dependency graph")?;

        // Execute level by level
        let mut results = HashMap::new();
        let mut completed = HashSet::new();
        let mut failed = HashSet::new();
        let mut max_concurrency = 0;

        let executor = Arc::new(executor);

        for (level_idx, level_tasks) in graph.levels.iter().enumerate() {
            // Track max concurrency
            max_concurrency = max_concurrency.max(level_tasks.len());

            // Execute all tasks in this level concurrently
            let level_results = self
                .execute_level(level_idx, level_tasks, executor.clone())
                .await?;

            // Process results
            for result in level_results {
                if result.success {
                    completed.insert(result.id.clone());
                } else {
                    failed.insert(result.id.clone());
                }
                results.insert(result.id.clone(), result);
            }

            // Stop if failure and not continuing
            if !failed.is_empty() && !self.config.continue_on_failure {
                break;
            }
        }

        let total_duration_ms = start_time.elapsed().as_millis() as u64;

        Ok(ParallelExecutionResult {
            success: failed.is_empty(),
            results,
            total_duration_ms,
            max_concurrency_used: max_concurrency,
            completed_count: completed.len(),
            failed_count: failed.len(),
        })
    }

    /// Execute all tasks in a single level concurrently
    async fn execute_level<F, Fut>(
        &self,
        level: usize,
        task_ids: &[String],
        executor: Arc<F>,
    ) -> Result<Vec<SubtaskResult>>
    where
        F: Fn(String) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<String>> + Send + 'static,
    {
        let semaphore = Arc::new(Semaphore::new(self.config.max_concurrent));
        let mut handles: Vec<JoinHandle<SubtaskResult>> = Vec::new();

        for task_id in task_ids {
            let task_id = task_id.clone();
            let executor = executor.clone();
            let semaphore = semaphore.clone();
            let timeout_secs = self.config.task_timeout_secs;

            let handle = tokio::spawn(async move {
                // Acquire semaphore permit
                let _permit = semaphore.acquire().await.unwrap();

                let start = std::time::Instant::now();
                let result = if timeout_secs > 0 {
                    // Execute with timeout
                    match tokio::time::timeout(
                        tokio::time::Duration::from_secs(timeout_secs),
                        executor(task_id.clone()),
                    )
                    .await
                    {
                        Ok(Ok(output)) => Ok(output),
                        Ok(Err(e)) => Err(e),
                        Err(_) => Err(anyhow::anyhow!(
                            "Task timed out after {} seconds",
                            timeout_secs
                        )),
                    }
                } else {
                    // Execute without timeout
                    executor(task_id.clone()).await
                };

                let duration_ms = start.elapsed().as_millis() as u64;

                match result {
                    Ok(output) => SubtaskResult {
                        id: task_id,
                        success: true,
                        level,
                        duration_ms,
                        error: None,
                        output: Some(output),
                    },
                    Err(e) => SubtaskResult {
                        id: task_id,
                        success: false,
                        level,
                        duration_ms,
                        error: Some(e.to_string()),
                        output: None,
                    },
                }
            });

            handles.push(handle);
        }

        // Wait for all tasks to complete
        let mut results = Vec::new();
        for handle in handles {
            let result = handle.await.context("Task panicked")?;
            results.push(result);
        }

        Ok(results)
    }

    /// Get configuration
    pub fn config(&self) -> &ParallelExecutorConfig {
        &self.config
    }
}

// ============================================================================
// Tests
// ============================================================================

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicUsize, Ordering};
    use tokio::time::{Duration, sleep};

    #[tokio::test]
    async fn test_execute_parallel_tasks() {
        let executor = ParallelExecutor::with_defaults();

        let subtasks = vec![
            ("A".to_string(), vec![]),
            ("B".to_string(), vec![]),
            ("C".to_string(), vec![]),
        ];

        let result = executor
            .execute(subtasks, |id| async move {
                sleep(Duration::from_millis(10)).await;
                Ok(format!("Completed {}", id))
            })
            .await
            .unwrap();

        assert!(result.success);
        assert_eq!(result.completed_count, 3);
        assert_eq!(result.failed_count, 0);
    }

    #[tokio::test]
    async fn test_execute_with_dependencies() {
        let executor = ParallelExecutor::with_defaults();

        // A -> B -> C (linear chain)
        let subtasks = vec![
            ("A".to_string(), vec![]),
            ("B".to_string(), vec!["A".to_string()]),
            ("C".to_string(), vec!["B".to_string()]),
        ];

        let order = Arc::new(AtomicUsize::new(0));

        let result = executor
            .execute(subtasks, {
                let order = order.clone();
                move |id| {
                    let order = order.clone();
                    async move {
                        let seq = order.fetch_add(1, Ordering::SeqCst);
                        sleep(Duration::from_millis(10)).await;
                        Ok(format!("{} executed at {}", id, seq))
                    }
                }
            })
            .await
            .unwrap();

        assert!(result.success);
        assert_eq!(result.completed_count, 3);

        // Check that B's level is higher than A's
        let a_level = result.results.get("A").unwrap().level;
        let b_level = result.results.get("B").unwrap().level;
        let c_level = result.results.get("C").unwrap().level;
        assert!(a_level < b_level);
        assert!(b_level < c_level);
    }

    #[tokio::test]
    async fn test_stop_on_failure() {
        let config = ParallelExecutorConfig {
            continue_on_failure: false,
            ..Default::default()
        };
        let executor = ParallelExecutor::new(config);

        let subtasks = vec![
            ("A".to_string(), vec![]),
            ("B".to_string(), vec![]),
            ("C".to_string(), vec!["A".to_string(), "B".to_string()]),
        ];

        let result = executor
            .execute(subtasks, |id| async move {
                if id == "B" {
                    Err(anyhow::anyhow!("Task B failed"))
                } else {
                    Ok(format!("Completed {}", id))
                }
            })
            .await
            .unwrap();

        assert!(!result.success);
        assert_eq!(result.failed_count, 1);
        // C should not execute because B failed
        assert!(!result.results.contains_key("C"));
    }

    #[tokio::test]
    async fn test_continue_on_failure() {
        let config = ParallelExecutorConfig {
            continue_on_failure: true,
            ..Default::default()
        };
        let executor = ParallelExecutor::new(config);

        let subtasks = vec![
            ("A".to_string(), vec![]),
            ("B".to_string(), vec![]),
            ("C".to_string(), vec!["A".to_string()]),
        ];

        let result = executor
            .execute(subtasks, |id| async move {
                if id == "B" {
                    Err(anyhow::anyhow!("Task B failed"))
                } else {
                    Ok(format!("Completed {}", id))
                }
            })
            .await
            .unwrap();

        assert!(!result.success); // Overall failed
        assert_eq!(result.failed_count, 1);
        assert_eq!(result.completed_count, 2);
        // C should execute because we continue on failure
        assert!(result.results.contains_key("C"));
    }

    #[tokio::test]
    async fn test_concurrency_limit() {
        let config = ParallelExecutorConfig {
            max_concurrent: 2,
            ..Default::default()
        };
        let executor = ParallelExecutor::new(config);

        let active = Arc::new(AtomicUsize::new(0));
        let max_active = Arc::new(AtomicUsize::new(0));

        let subtasks = vec![
            ("A".to_string(), vec![]),
            ("B".to_string(), vec![]),
            ("C".to_string(), vec![]),
            ("D".to_string(), vec![]),
        ];

        let result = executor
            .execute(subtasks, {
                let active = active.clone();
                let max_active = max_active.clone();
                move |id| {
                    let active = active.clone();
                    let max_active = max_active.clone();
                    async move {
                        let current = active.fetch_add(1, Ordering::SeqCst) + 1;
                        max_active.fetch_max(current, Ordering::SeqCst);

                        sleep(Duration::from_millis(50)).await;

                        active.fetch_sub(1, Ordering::SeqCst);
                        Ok(format!("Completed {}", id))
                    }
                }
            })
            .await
            .unwrap();

        assert!(result.success);
        // Max concurrency should not exceed configured limit
        let max = max_active.load(Ordering::SeqCst);
        assert!(max <= 2, "Max concurrency {} exceeded limit 2", max);
    }

    #[tokio::test]
    async fn test_task_timeout() {
        let config = ParallelExecutorConfig {
            task_timeout_secs: 1,
            ..Default::default()
        };
        let executor = ParallelExecutor::new(config);

        let subtasks = vec![("A".to_string(), vec![])];

        let result = executor
            .execute(subtasks, |_id| async move {
                // Sleep longer than timeout
                sleep(Duration::from_secs(2)).await;
                Ok("Should not complete".to_string())
            })
            .await
            .unwrap();

        assert!(!result.success);
        assert_eq!(result.failed_count, 1);

        let a_result = result.results.get("A").unwrap();
        assert!(!a_result.success);
        assert!(a_result.error.as_ref().unwrap().contains("timed out"));
    }
}
